### PR TITLE
Add a fuzzy search on tilelayers

### PR DESCRIPTION
**Pull request by HaudinFlorence**
Propose a version of the geojson renderer including a fuzzy search that compares a user input to the items of the list of tilelayers (from xyz services). Results are proposed in a dropdown menu as shown in the following gif :

fuzzy_search7

.

### CODE REVIEW

File: packages/geojson-extension/src/index.ts [hunk 1]
```diff
@@ -16,9 +22,34 @@ import 'leaflet/dist/leaflet.css';
...
+
+import * as access_d from './access_data.json';
+
+const tilelayers_data: { [key: string]: any } = providers;
+const access_data: { [key: string]: any } = access_d;
+const nameList: Array<string> = [];
+for (const [key1, val1] of Object.entries(tilelayers_data)) {
+  if (Object.keys(val1).includes('url')) {
+    const name = tilelayers_data[key1].name;
+    if (name !== undefined && nameList.includes(name) === false) {
```
Comment (davidbrochart) [inline packages/geojson-extension/src/index.ts:40]:
```suggestion
    if (name !== undefined && !nameList.includes(name)) {
```

File: packages/geojson-extension/src/index.ts [hunk 2]
```diff
@@ -16,9 +22,34 @@ import 'leaflet/dist/leaflet.css';
...
+      nameList.push(name);
+    }
+  } else {
+    for (const key2 of Object.keys(val1)) {
+      const name = tilelayers_data[key1][key2].name;
+      if (name !== undefined && nameList.includes(name) === false) {
```
Comment (davidbrochart) [inline packages/geojson-extension/src/index.ts:46]:
```suggestion
      if (name !== undefined && !nameList.includes(name)) {
```

File: packages/geojson-extension/src/index.ts [hunk 3]
```diff
@@ -51,22 +82,194 @@ leaflet.Icon.Default.mergeOptions({
...
+   */
+  readonly value: string;
+
+  /**
+   * The indices of the matched characters.
+   */
+  readonly indices: ReadonlyArray<number> | null;
+
+  /**
+   * The indices of the matched characters.
```
Comment (davidbrochart) [inline packages/geojson-extension/src/index.ts:97]: This description doesn't seem to match.

File: packages/geojson-extension/src/index.ts [hunk 4]
```diff
@@ -100,15 +303,77 @@ export class RenderedGeoJSON extends Widget implements IRenderMime.IRenderer {
...
+        marginRight: '0px',
+      });
+
+      button.onclick = () =>
+        showDialog({
+          title: '',
+          body: new TilelayerPalette(nameList),
+          buttons: [Dialog.cancelButton(), Dialog.okButton()],
+        }).then((result) => {
+          console.log('result.value: ', result.value);
```
Comment (davidbrochart) [inline packages/geojson-extension/src/index.ts:324]: Leftover?

**Review by davidbrochart — CHANGES_REQUESTED** [review]
You don't seem to use `packages/geojson-extension/style/icons/layers-16px.svg`, are you sure you want to commit it?

**Comment by davidbrochart** [comment]
I would be nice to have the fuzzy search show all results before entering some text, so that we know what base maps are available. Currently it shows nothing until we start typing.

**Comment by davidbrochart** [comment]
Also, maybe display the base map for OpenStreetMap by default?

**Comment by HaudinFlorence** [comment]
> I would be nice to have the fuzzy search show all results before entering some text, so that we know what base maps are available. Currently it shows nothing until we start typing.

I totally agree. I tried to do it but without any success. I tried to display the select list prior to use the eventhandler and to perform any fuzzy search, but the results was not as expected : I could not see more that one element of the list even when asking the size to be a given number and the css style was not applied.

**Comment by HaudinFlorence** [comment]
> Also, maybe display the base map for OpenStreetMap by default?

That's possible.

File: packages/geojson-extension/src/access_data.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{
+  "Thunderforest": {
+    "keyString": "apikey"
+  },
```
Comment (davidbrochart) [inline packages/geojson-extension/src/access_data.json:4]: Then maybe no need for `"keyString"`?
```suggestion
  "Thunderforest": "apikey",
```

File: packages/geojson-extension/src/index.ts [hunk 6]
```diff
@@ -100,15 +308,76 @@ export class RenderedGeoJSON extends Widget implements IRenderMime.IRenderer {
...
+        }).then((result) => {
+          const input_name = result.value;
+
+          if (input_name.includes('.')) {
+            const APIname = input_name.split('.')[0];
+            const subname = input_name.split('.')[1];
+            if (Object.keys(access_data).includes(APIname)) {
+              showDialog({
+                title: '',
+                body: new TextInput('Enter the APIkey please'),
```
Comment (davidbrochart) [inline packages/geojson-extension/src/index.ts:337]:
```suggestion
                body: new TextInput('Enter the API key please'),
```

**Review by davidbrochart — APPROVED** [review]
Thanks for the changes @HaudinFlorence.

A couple of potential improvements:
- When the API key is wrong, nothing happens. Maybe we could catch the errors and notify the user?

https://user-images.githubusercontent.com/4711805/177133822-696e5057-c46d-425b-bd1f-be7270286bae.mp4

- Some maps are not available, maybe worth removing them?

https://user-images.githubusercontent.com/4711805/177133838-8095a1ea-4fec-425a-89c4-ddd6af200550.mp4

**Comment by davidbrochart** [comment]
> I would be nice to have the fuzzy search show all results before entering some text, so that we know what base maps are available. Currently it shows nothing until we start typing.

Should be fixed in https://github.com/HaudinFlorence/jupyter-renderers/pull/2.

**Comment by HaudinFlorence** [comment]
> Thanks for the changes @HaudinFlorence.
> 
> A couple of potential improvements:
> 
> * When the API key is wrong, nothing happens. Maybe we could catch the errors and notify the user?
> 
>  georenderer.mp4 
> * Some maps are not available, maybe worth removing them?
> 
>  georenderer2.mp4

Thanks for the suggestions. It would be indeed better to have an error message when the API key is wrong. It would also be worthy to remove the tilelayers that are not available.

File: packages/geojson-extension/src/access_data.json (hunk 5 above)
Comment (HaudinFlorence) [reply packages/geojson-extension/src/access_data.json]: keyString has been removed.

File: packages/geojson-extension/src/index.ts (hunk 6 above)
Comment (HaudinFlorence) [reply packages/geojson-extension/src/index.ts]: This has been changed.

**Comment by HaudinFlorence** [comment]
> > I would be nice to have the fuzzy search show all results before entering some text, so that we know what base maps are available. Currently it shows nothing until we start typing.
> 
> Should be fixed in [HaudinFlorence#2](https://github.com/HaudinFlorence/jupyter-renderers/pull/2).

Thanks for the suggestion. This will be added.

File: packages/geojson-extension/src/index.ts [hunk 7]
```diff
@@ -100,15 +321,76 @@ export class RenderedGeoJSON extends Widget implements IRenderMime.IRenderer {
...
+                  tilelayers_data[APIname][subname]
+                );
+                layer.addTo(this._map);
+              });
+            } else {
+              const layer = leaflet.tileLayer(
+                tilelayers_data[APIname][subname].url,
+                tilelayers_data[APIname][subname]
+              );
+              layer.addTo(this._map);
```
Comment (davidbrochart) [inline packages/geojson-extension/src/index.ts:366]: It looks like layers are never removed from the map, so if the user chooses a base map and then another one, (unnecessary) requests will be made to the previous base map.

File: packages/geojson-extension/src/index.ts [hunk 8]
```diff
@@ -16,9 +20,40 @@ import 'leaflet/dist/leaflet.css';
...
+      nameList.push(name);
+    }
+  } else {
+    for (const key2 of Object.keys(val1)) {
+      const name = tilelayers_data[key1][key2].name;
+      if (name !== undefined && !nameList.includes(name)) {
+        if (
+          name !== 'OpenStreetMap.BlackAndWhite' &&
+          !name.includes('HikeBike') &&
+          !name.includes('HERE.')
```
Comment (davidbrochart) [inline packages/geojson-extension/src/index.ts:48]: Should we directly remove these base maps from `providers.json`?

File: packages/geojson-extension/src/index.ts (hunk 7 above)
Comment (HaudinFlorence) [reply packages/geojson-extension/src/index.ts:366]: > It looks like layers are never removed from the map, so if the user chooses a base map and then another one, (unnecessary) requests will be made to the previous base map.

Yes the layers are not removed the way the code is written now. Do you suggest to remove the previous basemap before trying to add a new one ?

File: packages/geojson-extension/src/index.ts [hunk 9]
```diff
@@ -100,15 +321,86 @@ export class RenderedGeoJSON extends Widget implements IRenderMime.IRenderer {
...
+            const APIname = input_name;
+            if (access_data[APIname] !== undefined) {
+              showDialog({
+                title: '',
+                body: new TextInput('Enter the API key please'),
+                buttons: [Dialog.cancelButton(), Dialog.okButton()],
+              }).then((result) => {
+                const APIkey = access_data[APIname];
+                tilelayers_data[APIname][APIkey] = result.value;
+                console.log('url :', tilelayers_data[APIname].url);
```
Comment (davidbrochart) [inline packages/geojson-extension/src/index.ts:383]: Leftover?

File: packages/geojson-extension/src/index.ts [hunk 10]
```diff
@@ -51,22 +86,203 @@ leaflet.Icon.Default.mergeOptions({
...
+  implements Dialog.IBodyWidget<string>
+{
+  constructor(list: Array<string> = []) {
+    super();
+    this.node.style.height = '270px';
+    this.node.style.height = '400px';
+    this._query = document.createElement('input');
+    this._query.type = 'text';
+    this._query.style.width = '400px';
+    this._query.placeholder = 'Enter your query';
```
Comment (jtpio) [inline packages/geojson-extension/src/index.ts:233]: Maybe something like `Search for a tile layer` would be more specific?

File: packages/geojson-extension/src/index.ts [hunk 11]
```diff
@@ -51,22 +86,203 @@ leaflet.Icon.Default.mergeOptions({
...
+
+    // Add the score to the results.
+    scores.push(score);
+  }
+
+  // Return the final array of scores.
+  return scores;
+}
+
+export class TilelayerPalette
```
Comment (jtpio) [inline packages/geojson-extension/src/index.ts:222]: Looks like this one is usually spelled `TileLayer` with a capital `L`, for example in Leaflet. Maybe we can keep the same style here and in other places?

File: packages/geojson-extension/style/base.css [hunk 12]
```diff
@@ -29,3 +29,30 @@
...
+}
+
+.input-content {
+  font-family: 'Arial';
+  font-size: 12px;
+  /*background-color: rgb(53, 111, 220);*/
+  color: black;
+}
+
+.select-list {
```
Comment (jtpio) [inline packages/geojson-extension/style/base.css:54]: Wondering if these classes could be more narrowly scoped to only apply to the geojson view, for example `.jp-RenderedGeoJSON`.

**Comment by jtpio** [comment]
Thanks @HaudinFlorence for working on this :+1: 

This might be similar to the comment above, but it looks like some layers don't prompt for an API key. For example:

https://user-images.githubusercontent.com/591645/177789721-4fe66f40-2671-43c0-88f1-36b32d014ff3.mp4

File: packages/geojson-extension/src/index.ts (hunk 10 above)
Comment (HaudinFlorence) [reply packages/geojson-extension/src/index.ts]: Thanks for the comment. This has been changed.

**Comment by HaudinFlorence** [comment]
> Thanks @HaudinFlorence for working on this +1
> 
> This might be similar to the comment above, but it looks like some layers don't prompt for an API key. For example:
> 
>  geojson-layers.mp4

I checked precisely for this tile layer and it seems to work from side, at least locally.

**Comment by HaudinFlorence** [comment]
> Thanks @HaudinFlorence for working on this +1
> 
> This might be similar to the comment above, but it looks like some layers don't prompt for an API key. For example:
> 
>  geojson-layers.mp4

Thank you very much for your comments and review :+1: @davidbrochart @jtpio

File: packages/geojson-extension/src/index.ts (hunk 9 above)
Comment (HaudinFlorence) [reply packages/geojson-extension/src/index.ts]: This has been addressed. Thank you very much for your review @davidbrochart

File: packages/geojson-extension/style/icons/layers-32px.svg [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+   transform="matrix(0.128,0,0,0.128,-229.6903,-39.078004)">
+</g>
+<g
+   id="g114"
+   transform="matrix(0.128,0,0,0.128,-229.6903,-39.078004)">
+</g>
+<g
+   id="g116"
+   transform="matrix(0.128,0,0,0.128,-229.6903,-39.078004)">
+</g>
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply packages/geojson-extension/style/icons/layers-32px.svg:102]: > Hint when using Inkscape, you can use `Save As -> Optimized SVG` to get a first lighter file. Then I finish to clean up the file manually.

File: packages/geojson-extension/package.json [hunk 14]
```diff
@@ -29,7 +29,10 @@
     "lib/*.d.ts",
     "lib/*.js",
     "style/*.*",
```
Comment (fcollonval) [inline packages/geojson-extension/package.json:31]: That should be enough to include everything in the `style` folder:

```suggestion
    "style/**/*",
```

File: packages/geojson-extension/package.json [hunk 15]
```diff
@@ -29,7 +29,10 @@
...
-    "style/index.js"
+    "styles/icons/*.svg",
+    "style/index.js",
+    "src/access_data.json",
+    "src/providers.json"
```
Comment (fcollonval) [inline packages/geojson-extension/package.json:35]:
```suggestion
    "src/*.json"
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
Minor comments on the packaging

File: packages/geojson-extension/style/icons/layers-32px.svg (hunk 13 above)
[the remark this answers is not in the document]
Comment (HaudinFlorence) [reply packages/geojson-extension/style/icons/layers-32px.svg]: Thanks for the comment and hint. A new version of the svg file is proposed in commit f763dea, generated with `Save As -> Optimized SVG` using Inkscape.

**Review by fcollonval — APPROVED** [review]
Thanks @HaudinFlorence

**Comment by HaudinFlorence** [comment]
Thanks for merging merging @fcollonval

**Pull request merged**